Shareable Retro Templates

mattkrick opened the pull request:
This introduces sharable templates.

### Scopes
Now, a template has a "scope" which can be
- team scope (the template is only visible to your team)
- org scope (the template is only visible to your org)
- public scope (the template is visible to the world)

### Ownership
When a user creates a team, we no longer create 5 starter templates at the team scope.
Instead, we show them their org & public templates & if they want to modify them, they can "clone" them (clone is used because it's less techy than "fork")
Cloning a template makes a copy of it & sets the scope to team
Visually, the template that was cloned should have a flag on it showing that it was cloned

### Ordering
templates are grouped by scope to make it fast to find what they want
team templates have a hard limit of 20
org & public templates are paginated because there could be lots
Templates are ordered based on hotness
hotness is a function of how old the template is & how often it gets used
for team & org-scoped templates, we put more emphasis on how new the template is, we call this the `newHotnessFactor`
for public templates, we put more emphasis on how many meetings have been run using that template
Both template age & number of meetings run follow the same S-shape curve. This is because the standard decay function (negative exponential distribution) works for things that age fast. templates only get used 2x/month, so we gotta slow it down.

- If a template is brand new, it gets 1 point. if it's 90+ days old, it gets 0. if it's 30 days old, it gets _almost_ 1 point. if it's 60 days old it gets _almost_ 0 (hence the S-shape)
- For each meeting that has been run using that template, if the meeting was run 90+ days ago, that gives it _almost_ 0 points. if it was run fewer than 30 days ago, it gets _almost_ 1 point
- the final score is the sumproduct of `newHotnessFactor * templateAge + (1-newHotnessFactor) * meetingAges`
- in order for a template to make it to the public templates list, it must have been used 5 times across 3+ teams

### List inclusion
If your template is on the public scope, does it belong in your team template list, the organization list, and/or the public list?
- My initial thought is to only have it appear at the lowest scope, ie the team scope. since you own it, this is where you'd expect to find it, so it makes sense.
- if another team in my org makes their template public, i should look for it in the org section. we can gamify with tags if they are super famous public ones. 

TEST
- [x] all team templates that have never been used are deleted
- [x] "public templates" shows all the starter templates
- [x] if no team templates exist, the empty state has a link to public templates
- [x] each scope is separated by tabs & should be fetched independently (prefetching on hover is OK).  
- [x]  "team templates" shows all templates that have been created or cloned from org or public templates
- [x] "org templates" has an empty state that says no one in their organization has shared a template
- [x] if "team templates" is empty, there is an empty state to encourage cloning a public template
- [x] when a new team is created, they have empty team & org templates
- [x] can have 20 team templates & 25 org templates & 5 public templates. it looks OK
- [x] team & public templates count towards the 20 template total
- [x] switching to a new scope selects the top template in that view. if no templates are available, the existing selected template stays the same
- [x] can clone an org or public template & it appears in the team templates. the view redirects to show the new template.
- [x] higher scopes filter out templates that exist in lower scopes (ie org templates do not include any team templates). for gamification, we'll use badges later
- [x] all templates show you the currently scope that they're at
- [x] team templates allow you to change the scope if you're the team lead
- [x] if your team template scope is decreased & the template has been used by another team, then a clone is created & the original is inactivated so it doesn't show up in that scope anymore (but past meetings can still reference it). 
- [x] removing a template selects the next closest template as the active one. if no other templates are left in that view, it moves to the public view
- [x] if another team removes their org template & i'm a part of that org but on a different team, i see the template go away
- [x] cloning the same template twice results in a suffix of `#x` where x is a number from 2 to 19.
- [x] can change the colors of the team template prompts
- [x] on uneditable prompts (org, public) they don't look clickable
- [x] a template shared publically by a team in your org shows up in the org list
- [x] Team A creates a template, shares it with the Org. Team B uses that template. Team A downscopes that template. Team B starts another meeting using that old template. Team B can create & edit reflections in that column

mattkrick commented:
hey @ackernaut can you give this a quick pass? desktop only. i think i may have just been staring at this for faaar too long

ackernaut commented:
Sure thing, taking a card…

ackernaut commented:
### Test cycle no. 1
- [ ] all team templates that have never been used are deleted
- [ ] "public templates" shows all the starter templates
- [ ] if no team templates exist, the empty state has a link to public templates
- [ ] each scope is separated by tabs & should be fetched independently (prefetching on hover is OK).  
- [ ] "team templates" shows all templates that have been created or cloned from org or public templates
- [ ] "org templates" has an empty state that says no one in their organization has shared a template
- [ ] if "team templates" is empty, there is an empty state to encourage cloning a public template
- [ ] when a new team is created, they have empty team & org templates
- [ ] can have 20 team templates & 25 org templates & 5 public templates. it looks OK
- [ ] team & public templates count towards the 20 template total
- [ ] switching to a new scope selects the top template in that view. if no templates are available, the existing selected template stays the same
- [ ] can clone an org or public template & it appears in the team templates. the view redirects to show the new template.
- [ ] higher scopes filter out templates that exist in lower scopes (ie org templates do not include any team templates). for gamification, we'll use badges later
- [ ] all templates show you the currently scope that they're at
- [ ] team templates allow you to change the scope if you're the team lead
- [ ] if your team template scope is decreased & the template has been used by another team, then a clone is created & the original is inactivated so it doesn't show up in that scope anymore (but past meetings can still reference it). 
- [ ] removing a template selects the next closest template as the active one. if no other templates are left in that view, it moves to the public view
- [ ] if another team removes their org template & i'm a part of that org but on a different team, i see the template go away
- [ ] cloning the same template twice results in a suffix of `#x` where x is a number from 2 to 19.
- [ ] can change the colors of the team template prompts
- [ ] on uneditable prompts (org, public) they don't look clickable
- [ ] a template shared publically by a team in your org shows up in the org list
- [ ] Team A creates a template, shares it with the Org. Team B uses that template. Team A downscopes that template. Team B starts another meeting using that old template. Team B can create & edit reflections in that column

ackernaut commented:
@mattkrick here’s what I got so far:
### The main issues

>if your team template scope is decreased & the template has been used by another team, then a clone is created & the original is inactivated so it doesn't show up in that scope anymore (but past meetings can still reference it).

>if another team removes their org template & i'm a part of that org but on a different team, i see the template go away

>Team A creates a template, shares it with the Org. Team B uses that template. Team A down-scopes that template. Team B starts another meeting using that old template. Team B can create & edit reflections in that column

- I’m not sure that the template list is refreshing in real-time for other teams on the org when a template is down-scoped by the owning team
- There can be a strange selection issue when trying to select a template that shows in the org list but was actually downloaded
- I get an error when trying to create reflections on a copy of the down-scoped issues
- It takes a hard refresh at any point using the app to get the list to refresh the next time it’s viewed

### Minor
- I haven’t counted the 20 template limit yet, but I’m not that worried about that
- I’m making a small list of style fixes, and I’ll try to make that in a single commit (the uneditable prompts can be clicked and have a dark outline so I didn’t check that test yet)

ackernaut commented:
@mattkrick I made a first pass at styling. If you pull down, let me know what you think.

ackernaut commented:
Okay, I pushed some more styles and [updated my comment above to reflect latest](https://github.com/ParabolInc/parabol/pull/3942#issuecomment-673597198)

ackernaut commented:
It would be nice to use our specific illustrations on our 5 public templates, and then use the custom one for team/org-owned templates.

ackernaut commented:
We could rename `Liked, Learned, Lacked, Longed for` to `Four L’s`?

ackernaut commented:
Would it bork the migration to use commas in `Start, Stop, Continue`?

mattkrick commented:
yeah, those should all be fine!

mattkrick commented:
> I’m not sure that the template list is refreshing in real-time for other teams on the org when a template is down-scoped by the owning team

Fixed! Note that things like prompts (color, description, name) will not update in real time if the subscriber is NOT on the same team as mutator. Doing this would require refactoring all those mutations on the org channel & i think it's just too much of an edge case

> There can be a strange selection issue when trying to select a template that shows in the org list but was actually downloaded

**need repro steps, i'm not quite sure i follow**

> I get an error when trying to create reflections on a copy of the down-scoped issues

fixed!

> It takes a hard refresh at any point using the app to get the list to refresh the next time it’s viewed

**I think fixing the first issue might have fixed this. could you retest & add repro if not?**

> Sometimes the description is Last used Just now — the capitalized ‘Just’ in the middle of the phrase is weird

fixed!

> When tapping to add a new template from the Public scope tab, it doesn’t load the new template details

fixed!

ackernaut commented:
Here’s a strange one @mattkrick 
- Go to the Public tab
- Tap each template from the top to the bottom
- When you tap Glad, Sad, Mad it flashes and goes back to Start, Stop, Continue
- I tried cloning and making 5 templates and I can click and select all 5 of the Team templates

ackernaut commented:
When I clone ‘Liked, Learned, Lacked, Longed for’ then it says ‘Four L’s Copy’. I tried migrating down and back up.

mattkrick commented:
those both smell like migration issues, if it was a successful down/up migration `‘Liked, Learned, Lacked, Longed for` shouldn't exist anymore in the Public template list. i'll try to repro, but we should definitely test this hard on staging

mattkrick commented:
doh! i bet it was a caching issue.
i'll write a script that wipes the redis cache in development so this doesn't bite you again

mattkrick commented:
OK merging to master so we can get this out to staging & give it a good ol kick there

ackernaut commented on scripts/dev.js:
```diff
@@ -102,6 +103,9 @@ const dev = async (maybeInit, isDangerous) => {
   fork(path.join(PROJECT_ROOT, 'dev/gqlExecutor.js'))
+  const redis = new Redis(process.env.REDIS_URL)
+  // it's nice to flush the cache, but can comment this out if you want to test cache hits between restarts
+  redis.flushall()
```
Nice!
